packages/agent/src/connect-v2.ts

```diff
@@ -0,0 +1,199 @@
+  // TODO: use separate keys for ECDH vs. sign/verify. could maybe use secp256k1.
+  const clientDid = await DidDht.create();
+
+  // Hash the code verifier to use as a code challenge and to encrypt the Request Object.
+  const { codeVerifieru8a } = Oidc.generateRandomCodeVerifier();
+
+  // Derive the code challenge based on the code verifier
+  const { codeChallengeu8a, codeChallengeb64url } =
+    await Oidc.deriveCodeChallenge(codeVerifieru8a);
+
+  // build callback URL to pass into the auth request
+  const callbackEndpoint = Oidc.buildOidcUrl({
+    baseURL  : connectServerUrl,
+    endpoint : 'callback',
+  });
+
+  // build the PAR request
+  const request = await Oidc.createAuthRequest({
+    client_id             : callbackEndpoint,
+    scope                 : 'web5', // TODO: clear with frank
+    code_challenge        : codeChallengeb64url,
+    code_challenge_method : 'S256',
+    permission_requests   : permissionRequests,
+    redirect_uri          : callbackEndpoint,
+    // known customer credential defines these
+    client_metadata       : {
+      client_uri                     : clientUri,
+      subject_syntax_types_supported : ['did:dht'],
+    },
+  });
+
+  // Sign the Request Object using the Client DID's signing key.
+  const requestJwt = await Oidc.signJwt({
+    did  : clientDid,
+    data : request,
+  });
+
+  if (!requestJwt) {
+    throw new Error('Unable to sign requestObject');
+  }
+
+  // Encryption nonce
+  const nonce = utils.randomBytes(24);
+
+  // Encrypt the Request Object JWT using the code challenge.
+  const requestObjectJwe = await Oidc.encryptAuthRequest({
+    jwt           : requestJwt,
+    codeChallenge : codeChallengeu8a,
+    nonce,
+  });
+
+  // Convert the encrypted Request Object to URLSearchParams for form encoding.
+  const formEncodedRequest = new URLSearchParams({ request: requestObjectJwe });
+
+  const pushedAuthorizationRequestEndpoint = Oidc.buildOidcUrl({
+    baseURL  : connectServerUrl,
+    endpoint : 'pushedAuthorizationRequest',
+  });
+
+  try {
+    const parResponse = await fetch(pushedAuthorizationRequestEndpoint, {
+      body    : formEncodedRequest,
+      method  : 'POST',
+      headers : {
+        'Content-Type': 'application/x-www-form-urlencoded',
+      },
+    });
+    if (!parResponse.ok) {
+      throw new Error(`${parResponse.status}: ${parResponse.statusText}`);
+    }
+
+    const parData: PushedAuthResponse = await parResponse.json();
+
+    // a deeplink to a web5 compatible wallet. if the wallet scans this link it should receive
+    // a route to its web5 connect provider flow and the params of where to fetch the auth request.
+    const walletURI = new URL('web5://connect/');
+    walletURI.searchParams.set(
+      'nonce',
+      Convert.uint8Array(nonce).toBase64Url()
+    );
+    walletURI.searchParams.set('request_uri', parData.request_uri);
+    walletURI.searchParams.set('client_did', clientDid.uri);
```

Comment:
Sanity: did you mean `client_id`?

Reply:
no, I meant clientDid

Reply:
Sanity x2: not asking about `clientDid`, did you mean `client_did`? I don't see references to `client_did` in the context of SIOP with a quick search, but saw a tons of hits for `client_id`.

Are you able to send me a link on where `client_did` is described in spec? For my education. 🙏

Reply:
`client_did` is *our* param. it's part of web5 connect. the `client_id` that you're referring to is a website. it's described in the type definitions

Reply:
Sorry, sanity x3: I understand custom claims are fine, but isn't SIOPv2 using `client_id` for specifying the client DID?

e.g. https://openid.net/specs/openid-connect-self-issued-v2-1_0.html#section-7.2.3

So why invent another field for it?

Reply:
It's a field that accommodates either a DID or a callback URL. We have a callback URL. we already use 'client_id', it's just not in the query params. That being said I think we can get rid of the client did param altogether, as I already get it from the JWK

Reply:
Going to do more research on this, lets leave it open

Reply:
Ok you were right. We already got the callback endpoint from the redirect URI and we don't need to use `client_id` for it. Sorry, and nice catch!

Reply:
fixed

Reply:
Final FYI:

I see that you removed `client_id` from the URL, though, from examples that I can find in the specs on the internet, I always see `client_id` included in the URL showing that it is an exception to the rule when using `request_uri`. I am not going to insist to include `client_id` in the URL because you are right in saying that it is redundant, but it does "appear unconventional" to not include it. The rationale for having it in the URL seems so that it can act as a request lookup key/short-cut, but that's not how we implement `request_uri` generation.

You are good to resolve this comment thread if you see no further action on `client_id`.

Reply:
> I see that you removed client_id from the URL, though, from examples that I can find in the specs on the internet, I always see client_id included in the URL showing that it is an exception to the rule when using request_uri

We are using a PAR so we don't need to include it in query params, it's in the message body